File: packages/sdk/src/artifact/runtime.ts
import { Ajv2020 } from "ajv/dist/2020.js";
import type { AnySchema, ValidateFunction } from "ajv/dist/2020.js";
import { z } from "zod";
import manifestSchema from "./data-artifact-manifest.schema.json" with {
  type: "json",
};
import { blobKeySHA256Hex, Key } from "../datastore/key.ts";
import {
  type DatastoreClient,
  DatastoreConflictError,
  DatastoreNotFoundError,
  type ObjectInfo,
} from "../datastore/types.ts";
import {
  decodeCanonicalUtf8,
  type JsonValue,
  parseCanonicalJsonText,
  sha256RefBytes,
  stableStringify,
} from "../stable.ts";

export const JSON_CONTENT_TYPE = "application/json";
export const MANIFEST_CONTENT_TYPE =
  "application/vnd.massive.data-artifact-manifest+json";

export interface ArtifactDestination {
  readonly manifestKey: Key;
  readonly schema: string;
}

export interface ArtifactProducer {
  readonly projectKey: string;
  readonly planHash: string;
  readonly runId: string;
  readonly nodeId: string;
  readonly attempt: number;
  readonly scope?: ExecutionScope;
}

export interface MapItemScopeFrame {
  readonly kind: "map-item";
  readonly mapId: string;
  readonly index: number;
}

// Frames are ordered outer-to-inner. The optional wrapper leaves the static
// invocation identity unchanged while making nested map identities explicit.
export interface ExecutionScope {
  readonly frames: readonly MapItemScopeFrame[];
}

const hashRefSchema = z
  .string()
  .regex(/^sha256:[0-9a-f]{64}$/, "must be a lowercase SHA-256 reference");

const projectKeySchema = z
  .string()
  .regex(
    /^sha256-[0-9a-f]{64}$/,
    "must be a normalized SHA-256 project namespace key",
  );

const pathSegmentSchema = z
  .string()
  .min(1, "must not be empty")
  .max(128, "must be at most 128 characters")
  .regex(
    /^[A-Za-z0-9_.@:#-]+$/,
    "must contain only safe ASCII path-segment characters",
  )
  .refine((value) => value !== "." && value !== "..", {
    message: "must not be a dot path segment",
  });

const mapItemScopeFrameSchema = z.object({
  kind: z.literal("map-item"),
  mapId: pathSegmentSchema,
  index: z.number().int().min(0).max(Number.MAX_SAFE_INTEGER),
}).strict();

const executionScopeSchema = z.object({
  frames: z.array(mapItemScopeFrameSchema).min(1),
}).strict();

// The producer is the namespace identity for an immutable output slot. Keep
// its parsing in one schema rather than relying on Key.parse to accidentally
// reject only some unsafe values after work has already begun.
const artifactProducerSchema = z
  .object({
    projectKey: projectKeySchema,
    planHash: hashRefSchema,
    runId: pathSegmentSchema,
    nodeId: pathSegmentSchema,
    attempt: z
      .number()
      .int("must be an integer")
      .positive("must be positive")
      .max(Number.MAX_SAFE_INTEGER, "must be a safe integer"),
    scope: executionScopeSchema.optional(),
  })
  .strict();

const artifactDestinationSchema = z
  .object({
    manifestKey: z.custom<Key>((value): value is Key => value instanceof Key, {
      error: "must be a validated datastore key",
    }),
    schema: hashRefSchema,
  })
  .strict();

export interface ArtifactRef {
  readonly key: string;
  readonly hash: string;
  readonly size: number;
  readonly contentType: string;
}

export interface PublishedJson {
  readonly manifest: ArtifactRef;
  readonly body: ArtifactRef;
  readonly schema: string;
}

export interface ResolvedJson {
  readonly published: PublishedJson;
  readonly body: Uint8Array;
}

interface DataArtifactManifest {
  readonly kind: "DataArtifactManifest";
  readonly schemaVersion: 1;
  readonly encoding: "canonical-json-v0";
  readonly producer: ArtifactProducer;
  readonly schema: string;
  readonly body: ArtifactRef;
}

export class ArtifactError extends Error {
  constructor(message: string, options?: ErrorOptions) {
    super(message, options);
    this.name = "ArtifactError";
  }
}

export class ArtifactValidationError extends ArtifactError {
  constructor(message: string, options?: ErrorOptions) {
    super(message, options);
    this.name = "ArtifactValidationError";
  }
}

export class ArtifactIntegrityError extends ArtifactError {
  constructor(message: string, options?: ErrorOptions) {
    super(message, options);
    this.name = "ArtifactIntegrityError";
  }
}

export class ArtifactBodyConflictError extends ArtifactError {
  constructor(readonly key: Key) {
    super(`Artifact body conflict at ${key.toString()}`);
    this.name = "ArtifactBodyConflictError";
  }
}

export class ArtifactManifestConflictError extends ArtifactError {
  constructor(readonly key: Key) {
    super(`Artifact manifest conflict at ${key.toString()}`);
    this.name = "ArtifactManifestConflictError";
  }
}

export class ArtifactNotFoundError extends ArtifactError {
  constructor(readonly key: Key) {
    super(`Artifact object not found: ${key.toString()}`);
    this.name = "ArtifactNotFoundError";
  }
}

// ArtifactRuntime is deliberately the only seam that commits a produced JSON
// value. Callers provide the deterministic attempt destination and canonical
// value; this module owns body-first publication, immutable-manifest recovery,
// and all validation required before a value becomes visible downstream.
export class ArtifactRuntime {
  constructor(private readonly store: DatastoreClient) {}

  validateDestination(
    destination: ArtifactDestination,
    producer: ArtifactProducer,
  ): void {
    validateDestination(destination, validateProducer(producer));
  }

  async publishJson(
    destination: ArtifactDestination,
    producer: ArtifactProducer,
    body: string | Uint8Array,
  ): Promise<PublishedJson> {
    const validatedProducer = validateProducer(producer);
    validateDestination(destination, validatedProducer);
    const bytes = encode(body);
    await this.validateCanonicalJson(destination.schema, bytes);

    const bodyHash = sha256RefBytes(bytes);
    const bodyKey = keyForHash(bodyHash);
    const bodyRef: ArtifactRef = {
      key: bodyKey.toString(),
      hash: bodyHash,
      size: bytes.byteLength,
      contentType: JSON_CONTENT_TYPE,
    };
    const manifest: DataArtifactManifest = {
      kind: "DataArtifactManifest",
      schemaVersion: 1,
      encoding: "canonical-json-v0",
      producer: validatedProducer,
      schema: destination.schema,
      body: bodyRef,
    };
    const manifestBytes = await canonicalManifest(manifest);

    await putImmutable(
      this.store,
      bodyKey,
      bytes,
      JSON_CONTENT_TYPE,
      (key) => new ArtifactBodyConflictError(key),
    );
    await putImmutable(
      this.store,
      destination.manifestKey,
      manifestBytes,
      MANIFEST_CONTENT_TYPE,
      (key) => new ArtifactManifestConflictError(key),
    );

    return publishedJson(
      destination.manifestKey,
      manifestBytes,
      bodyRef,
      destination.schema,
    );
  }

  async resolveJson(
    destination: ArtifactDestination,
    producer: ArtifactProducer,
  ): Promise<ResolvedJson> {
    const validatedProducer = validateProducer(producer);
    validateDestination(destination, validatedProducer);
    const manifestObject = await getArtifactObject(
      this.store,
      destination.manifestKey,
    );
    if (manifestObject.info.contentType !== MANIFEST_CONTENT_TYPE) {
      throw new ArtifactIntegrityError(
        `Manifest ${destination.manifestKey.toString()} has content type ${manifestObject.info.contentType}`,
      );
    }

    const manifest = await decodeManifest(
      manifestObject.body,
      destination.manifestKey,
    );
    if (
      !sameProducer(manifest.producer, validatedProducer) ||
      manifest.schema !== destination.schema
    ) {
      throw new ArtifactIntegrityError(
        `Manifest ${destination.manifestKey.toString()} does not match its expected producer and schema`,
      );
    }

    const bodyKey = keyForHash(manifest.body.hash);
    if (bodyKey.toString() !== manifest.body.key) {
      throw new ArtifactIntegrityError(
        `Manifest ${destination.manifestKey.toString()} body key does not match its digest`,
      );
    }
    const bodyObject = await getArtifactObject(this.store, bodyKey);
    if (
      bodyObject.info.contentType !== JSON_CONTENT_TYPE ||
      bodyObject.body.byteLength !== manifest.body.size ||
      sha256RefBytes(bodyObject.body) !== manifest.body.hash
    ) {
      throw new ArtifactIntegrityError(
        `Body ${bodyKey.toString()} does not match its manifest`,
      );
    }
    try {
      await this.validateCanonicalJson(manifest.schema, bodyObject.body);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      throw new ArtifactIntegrityError(
        `Body ${bodyKey.toString()} is invalid: ${message}`,
        { cause: error },
      );
    }

    return {
      published: publishedJson(
        destination.manifestKey,
        manifestObject.body,
        manifest.body,
        manifest.schema,
      ),
      body: bodyObject.body,
    };
  }

  private async validateCanonicalJson(
    schemaRef: string,
    body: Uint8Array,
  ): Promise<void> {
    const value = parseCanonicalJson(body, "value");
    const schemaKey = keyForHash(schemaRef);
    let schemaObject;
    try {
      schemaObject = await this.store.get(schemaKey);
    } catch (error) {
      if (!(error instanceof DatastoreNotFoundError)) {
        throw error;
      }
      throw new ArtifactValidationError(
        `Cannot read schema ${schemaRef}: ${
          error instanceof Error ? error.message : String(error)
        }`,
        { cause: error },
      );
    }
    const schema = parseCanonicalJson(schemaObject.body, "schema");
    if (sha256RefBytes(schemaObject.body) !== schemaRef) {
      throw new ArtifactValidationError(
        `Schema ${schemaRef} is not stored under its digest`,
      );
    }

    let validate: ValidateFunction;
    try {
      validate = new Ajv2020({ allErrors: true, strict: true }).compile(
        schema as AnySchema,
      );
    } catch (error) {
      throw new ArtifactValidationError(
        `Schema ${schemaRef} could not be compiled: ${
          error instanceof Error ? error.message : String(error)
        }`,
        { cause: error },
      );
    }
    if (!validate(value)) {
      throw new ArtifactValidationError(
        `Value does not satisfy schema ${schemaRef}: ${
          formatValidationError(validate)
        }`,
      );
    }
  }
}

function validateDestination(
  destination: ArtifactDestination,
  producer: ArtifactProducer,
): void {
  const destinationResult = artifactDestinationSchema.safeParse(destination);
  if (!destinationResult.success) {
    throw new ArtifactValidationError(
      `Invalid artifact destination: ${
        z.prettifyError(destinationResult.error)
      }`,
      { cause: destinationResult.error },
    );
  }
  let expected: Key;
  try {
    expected = Key.parse(
      `projects/${producer.projectKey}/runs/${producer.runId}/steps/${producer.nodeId}${
        scopeKeySuffix(producer.scope)
      }/${producer.attempt}/output-manifest.json`,
    );
  } catch (error) {
    throw new ArtifactValidationError(
      `Invalid producer identity: ${
        error instanceof Error ? error.message : String(error)
      }`,
      { cause: error },
    );
  }
  if (destination.manifestKey.toString() !== expected.toString()) {
    throw new ArtifactValidationError(
      `Manifest destination ${destination.manifestKey.toString()} does not match producer slot ${expected.toString()}`,
    );
  }
}

function validateProducer(producer: unknown): ArtifactProducer {
  const result = artifactProducerSchema.safeParse(producer);
  if (!result.success) {
    throw new ArtifactValidationError(
      `Invalid producer identity: ${z.prettifyError(result.error)}`,
      { cause: result.error },
    );
  }
  return result.data;
}

async function putImmutable(
  store: DatastoreClient,
  key: Key,
  body: Uint8Array,
  contentType: string,
  conflict: (key: Key) => Error,
): Promise<ObjectInfo> {
  try {
    return await store.put(key, body, { contentType, ifAbsent: true });
  } catch (error) {
    if (!(error instanceof DatastoreConflictError)) {
      throw error;
    }
    const existing = await store.get(key);
    if (
      existing.info.contentType !== contentType ||
      !sameBytes(existing.body, body)
    ) {
      throw conflict(key);
    }
    return existing.info;
  }
}

async function canonicalManifest(
  manifest: DataArtifactManifest,
): Promise<Uint8Array> {
  const body = encode(stableStringify(manifest));
  try {
    await validateManifest(body);
  } catch (error) {
    throw new ArtifactValidationError(
      `Artifact manifest is invalid: ${
        error instanceof Error ? error.message : String(error)
      }`,
      { cause: error },
    );
  }
  return body;
}

async function decodeManifest(
  body: Uint8Array,
  key: Key,
): Promise<DataArtifactManifest> {
  let value: JsonValue;
  try {
    value = parseCanonicalJson(body, `manifest ${key.toString()}`);
  } catch (error) {
    throw new ArtifactIntegrityError(
      `Manifest ${key.toString()} is not canonical JSON: ${
        error instanceof Error ? error.message : String(error)
      }`,
      { cause: error },
    );
  }
  try {
    await validateManifest(body);
  } catch (error) {
    throw new ArtifactIntegrityError(
      `Manifest ${key.toString()} is invalid: ${
        error instanceof Error ? error.message : String(error)
      }`,
      { cause: error },
    );
  }
  return value as unknown as DataArtifactManifest;
}

let manifestValidator: Promise<ValidateFunction> | undefined;

function validateManifest(body: Uint8Array): Promise<void> {
  // Manifest JSON is also an immutable, hashed wire artifact. Parsing through
  // the canonical boundary before AJV prevents a second permissive JSON.parse
  // path from accepting whitespace or normalized numeric spellings.
  const value = parseCanonicalJson(body, "manifest");
  manifestValidator ??= compileManifestValidator();
  return manifestValidator.then((validate) => {
    if (!validate(value)) {
      throw new Error(formatValidationError(validate));
    }
  });
}

async function compileManifestValidator(): Promise<ValidateFunction> {
  return new Ajv2020({ allErrors: true, strict: true }).compile(
    manifestSchema as AnySchema,
  );
}

async function getArtifactObject(
  store: DatastoreClient,
  key: Key,
) {
  try {
    return await store.get(key);
  } catch (error) {
    if (error instanceof DatastoreNotFoundError) {
      throw new ArtifactNotFoundError(key);
    }
    throw error;
  }
}

function parseCanonicalJson(body: Uint8Array, role: string): JsonValue {
  try {
    return parseCanonicalJsonText(decodeCanonicalUtf8(body));
  } catch (error) {
    throw new ArtifactValidationError(
      `${role} is not canonical JSON: ${
        error instanceof Error ? error.message : String(error)
      }`,
      { cause: error },
    );
  }
}

function publishedJson(
  manifestKey: Key,
  manifestBody: Uint8Array,
  body: ArtifactRef,
  schema: string,
): PublishedJson {
  return {
    manifest: {
      key: manifestKey.toString(),
      hash: sha256RefBytes(manifestBody),
      size: manifestBody.byteLength,
      contentType: MANIFEST_CONTENT_TYPE,
    },
    body,
    schema,
  };
}

function keyForHash(hash: string): Key {
  if (!hash.startsWith("sha256:")) {
    throw new ArtifactValidationError(
      `Invalid SHA-256 reference ${JSON.stringify(hash)}`,
    );
  }
  try {
    return blobKeySHA256Hex(hash.slice("sha256:".length));
  } catch (error) {
    throw new ArtifactValidationError(
      `Invalid SHA-256 reference ${JSON.stringify(hash)}`,
      { cause: error },
    );
  }
}

function sameProducer(
  left: ArtifactProducer,
  right: ArtifactProducer,
): boolean {
  return left.projectKey === right.projectKey &&
    left.planHash === right.planHash &&
    left.runId === right.runId &&
    left.nodeId === right.nodeId &&
    left.attempt === right.attempt &&
    sameScope(left.scope, right.scope);
}

function sameScope(
  left: ExecutionScope | undefined,
  right: ExecutionScope | undefined,
): boolean {
  if (left === undefined || right === undefined) return left === right;
  return left.frames.length === right.frames.length &&
    left.frames.every((frame, index) => {
      const other = right.frames[index];
      return other !== undefined && frame.kind === other.kind &&
        frame.mapId === other.mapId && frame.index === other.index;
    });
}

function scopeKeySuffix(scope: ExecutionScope | undefined): string {
  if (scope === undefined) return "";
  return "/scopes" +
    scope.frames.map((frame) => `/maps/${frame.mapId}/items/${frame.index}`)
      .join("");
}

function sameBytes(left: Uint8Array, right: Uint8Array): boolean {
  if (left.byteLength !== right.byteLength) {
    return false;
  }
  for (let index = 0; index < left.byteLength; index += 1) {
    if (left[index] !== right[index]) {
      return false;
    }
  }
  return true;
}

function encode(body: string | Uint8Array): Uint8Array {
  return typeof body === "string" ? new TextEncoder().encode(body) : body;
}

function formatValidationError(validate: ValidateFunction): string {
  const error = validate.errors?.[0];
  if (error === undefined) {
    return "unknown JSON Schema violation";
  }
  const location = error.instancePath === "" ? "<root>" : error.instancePath;
  return `${location}: ${error.message ?? "is invalid"}`;
}

File: internal/artifact/runtime.go
// Package artifact publishes and resolves immutable canonical JSON values.
// A body is content-addressed first; its run-scoped manifest is the commit
// point that makes the value visible as a workflow output.
package artifact

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"sync"

	schemacontract "github.com/Sly1029/massive/conformance/schema"
	"github.com/Sly1029/massive/internal/canonical"
	"github.com/Sly1029/massive/internal/datastore"
	"github.com/santhosh-tekuri/jsonschema/v6"
)

const (
	JSONContentType     = "application/json"
	ManifestContentType = "application/vnd.massive.data-artifact-manifest+json"
	manifestSchemaRef   = "https://massive.dev/conformance/schema/data-artifact-manifest.schema.json"
)

var (
	ErrValidation       = errors.New("artifact validation failed")
	ErrIntegrity        = errors.New("artifact integrity check failed")
	ErrBodyConflict     = errors.New("artifact body conflict")
	ErrManifestConflict = errors.New("artifact manifest conflict")
)

var (
	safePathSegmentPattern = regexp.MustCompile(`^[A-Za-z0-9_.@:#-]+$`)
	safeProjectKeyPattern  = regexp.MustCompile(`^[A-Za-z0-9_.@:#-]+(/[A-Za-z0-9_.@:#-]+)*$`)
)

type Destination struct {
	ManifestKey datastore.Key
	Schema      string
}

type Producer struct {
	ProjectKey string `json:"projectKey"`
	PlanHash   string `json:"planHash"`
	RunID      string `json:"runId"`
	NodeID     string `json:"nodeId"`
	Attempt    int    `json:"attempt"`
}

type ArtifactRef struct {
	Key         string `json:"key"`
	Hash        string `json:"hash"`
	Size        int    `json:"size"`
	ContentType string `json:"contentType"`
}

type PublishedJSON struct {
	Manifest ArtifactRef
	Body     ArtifactRef
	Schema   string
}

type dataArtifactManifest struct {
	Kind          string      `json:"kind"`
	SchemaVersion uint32      `json:"schemaVersion"`
	Encoding      string      `json:"encoding"`
	Producer      Producer    `json:"producer"`
	Schema        string      `json:"schema"`
	Body          ArtifactRef `json:"body"`
}

// PublishJSON validates canonical JSON, convergently installs its content body,
// then conditionally creates the immutable manifest that commits visibility.
func PublishJSON(ctx context.Context, store datastore.Datastore, destination Destination, producer Producer, body []byte) (PublishedJSON, error) {
	if err := validateDestination(destination, producer); err != nil {
		return PublishedJSON{}, err
	}
	if err := validateCanonicalJSON(ctx, store, destination.Schema, body); err != nil {
		return PublishedJSON{}, err
	}

	bodyHash := canonical.DigestBytes(body)
	bodyKey, err := blobKey(bodyHash)
	if err != nil {
		return PublishedJSON{}, err
	}
	bodyRef := ArtifactRef{Key: bodyKey.String(), Hash: bodyHash, Size: len(body), ContentType: JSONContentType}
	manifest := dataArtifactManifest{
		Kind:          "DataArtifactManifest",
		SchemaVersion: 0,
		Encoding:      "canonical-json-v0",
		Producer:      producer,
		Schema:        destination.Schema,
		Body:          bodyRef,
	}
	manifestBytes, err := canonicalManifest(manifest)
	if err != nil {
		return PublishedJSON{}, err
	}

	if err := putImmutable(ctx, store, bodyKey, body, JSONContentType, ErrBodyConflict); err != nil {
		return PublishedJSON{}, err
	}
	if err := putImmutable(ctx, store, destination.ManifestKey, manifestBytes, ManifestContentType, ErrManifestConflict); err != nil {
		return PublishedJSON{}, err
	}

	return PublishedJSON{
		Manifest: ArtifactRef{
			Key:         destination.ManifestKey.String(),
			Hash:        canonical.DigestBytes(manifestBytes),
			Size:        len(manifestBytes),
			ContentType: ManifestContentType,
		},
		Body:   bodyRef,
		Schema: destination.Schema,
	}, nil
}

// ResolveJSON returns a value only after verifying its immutable manifest,
// producer, content-addressed body, canonical encoding, and pinned schema.
func ResolveJSON(ctx context.Context, store datastore.Datastore, destination Destination, producer Producer) (PublishedJSON, []byte, error) {
	if err := validateDestination(destination, producer); err != nil {
		return PublishedJSON{}, nil, err
	}
	manifestObject, err := store.Get(ctx, destination.ManifestKey)
	if err != nil {
		return PublishedJSON{}, nil, fmt.Errorf("read artifact manifest %s: %w", destination.ManifestKey, err)
	}
	if manifestObject.Info.ContentType != ManifestContentType {
		return PublishedJSON{}, nil, fmt.Errorf("%w: manifest %s content type is %q", ErrIntegrity, destination.ManifestKey, manifestObject.Info.ContentType)
	}
	canonicalBytes, err := canonical.CanonicalizeJSON(manifestObject.Body)
	if err != nil || !bytes.Equal(canonicalBytes, manifestObject.Body) {
		return PublishedJSON{}, nil, fmt.Errorf("%w: manifest %s is not canonical JSON", ErrIntegrity, destination.ManifestKey)
	}
	if err := validateManifestSchema(manifestObject.Body); err != nil {
		return PublishedJSON{}, nil, fmt.Errorf("%w: manifest %s: %v", ErrIntegrity, destination.ManifestKey, err)
	}

	var manifest dataArtifactManifest
	if err := json.Unmarshal(manifestObject.Body, &manifest); err != nil {
		return PublishedJSON{}, nil, fmt.Errorf("%w: decode manifest %s: %v", ErrIntegrity, destination.ManifestKey, err)
	}
	if manifest.Producer != producer || manifest.Schema != destination.Schema {
		return PublishedJSON{}, nil, fmt.Errorf("%w: manifest %s does not match its expected producer and schema", ErrIntegrity, destination.ManifestKey)
	}
	bodyKey, err := blobKey(manifest.Body.Hash)
	if err != nil || bodyKey.String() != manifest.Body.Key {
		return PublishedJSON{}, nil, fmt.Errorf("%w: manifest %s body key does not match its digest", ErrIntegrity, destination.ManifestKey)
	}
	bodyObject, err := store.Get(ctx, bodyKey)
	if err != nil {
		return PublishedJSON{}, nil, fmt.Errorf("%w: read manifest body %s: %v", ErrIntegrity, bodyKey, err)
	}
	if bodyObject.Info.ContentType != JSONContentType || len(bodyObject.Body) != manifest.Body.Size || canonical.DigestBytes(bodyObject.Body) != manifest.Body.Hash {
		return PublishedJSON{}, nil, fmt.Errorf("%w: body %s does not match its manifest", ErrIntegrity, bodyKey)
	}
	if err := validateCanonicalJSON(ctx, store, manifest.Schema, bodyObject.Body); err != nil {
		return PublishedJSON{}, nil, fmt.Errorf("%w: body %s: %v", ErrIntegrity, bodyKey, err)
	}

	return PublishedJSON{
		Manifest: ArtifactRef{Key: destination.ManifestKey.String(), Hash: canonical.DigestBytes(manifestObject.Body), Size: len(manifestObject.Body), ContentType: ManifestContentType},
		Body:     manifest.Body,
		Schema:   manifest.Schema,
	}, bodyObject.Body, nil
}

func validateDestination(destination Destination, producer Producer) error {
	if err := validateProducerIdentity(producer); err != nil {
		return err
	}
	want, err := datastore.ParseKey("projects/" + producer.ProjectKey + "/runs/" + producer.RunID + "/steps/" + producer.NodeID + "/" + strconv.Itoa(producer.Attempt) + "/output-manifest.json")
	if err != nil {
		return fmt.Errorf("%w: invalid producer identity: %v", ErrValidation, err)
	}
	if destination.ManifestKey != want {
		return fmt.Errorf("%w: manifest destination %s does not match producer slot %s", ErrValidation, destination.ManifestKey, want)
	}
	return nil
}

func validateProducerIdentity(producer Producer) error {
	if !isSafeProjectKey(producer.ProjectKey) {
		return fmt.Errorf("%w: project key %q is not a safe relative path", ErrValidation, producer.ProjectKey)
	}
	if !isSafePathSegment(producer.RunID) {
		return fmt.Errorf("%w: run ID %q is not a safe path segment", ErrValidation, producer.RunID)
	}
	if !isSafePathSegment(producer.NodeID) {
		return fmt.Errorf("%w: node ID %q is not a safe path segment", ErrValidation, producer.NodeID)
	}
	if producer.Attempt < 1 {
		return fmt.Errorf("%w: attempt must be at least one", ErrValidation)
	}
	return nil
}

func isSafePathSegment(value string) bool {
	return safePathSegmentPattern.MatchString(value) && value != "." && value != ".."
}

func isSafeProjectKey(value string) bool {
	if !safeProjectKeyPattern.MatchString(value) {
		return false
	}
	for _, segment := range strings.Split(value, "/") {
		if !isSafePathSegment(segment) {
			return false
		}
	}
	return true
}

func validateCanonicalJSON(ctx context.Context, store datastore.Datastore, schemaRef string, body []byte) error {
	canonicalBody, err := canonical.CanonicalizeJSON(body)
	if err != nil || !bytes.Equal(canonicalBody, body) {
		return fmt.Errorf("%w: value is not canonical JSON", ErrValidation)
	}
	schemaKey, err := blobKey(schemaRef)
	if err != nil {
		return fmt.Errorf("%w: invalid schema reference %q", ErrValidation, schemaRef)
	}
	schemaObject, err := store.Get(ctx, schemaKey)
	if err != nil {
		return fmt.Errorf("%w: read schema %s: %v", ErrValidation, schemaKey, err)
	}
	canonicalSchema, err := canonical.CanonicalizeJSON(schemaObject.Body)
	if err != nil || !bytes.Equal(canonicalSchema, schemaObject.Body) || canonical.DigestBytes(schemaObject.Body) != schemaRef {
		return fmt.Errorf("%w: schema %s is not canonical or does not match its digest", ErrValidation, schemaRef)
	}
	if err := validateJSONSchema(schemaObject.Body, body); err != nil {
		return fmt.Errorf("%w: value does not satisfy schema %s: %v", ErrValidation, schemaRef, err)
	}
	return nil
}

func putImmutable(ctx context.Context, store datastore.Datastore, key datastore.Key, body []byte, contentType string, conflict error) error {
	_, err := store.Put(ctx, key, body, datastore.PutOptions{ContentType: contentType, IfAbsent: true})
	if err == nil {
		return nil
	}
	if !errors.Is(err, datastore.ErrAlreadyExists) {
		return fmt.Errorf("publish immutable object %s: %w", key, err)
	}
	existing, getErr := store.Get(ctx, key)
	if getErr != nil {
		return fmt.Errorf("%w: read existing object %s after create conflict: %v", conflict, key, getErr)
	}
	if existing.Info.ContentType != contentType || !bytes.Equal(existing.Body, body) {
		return fmt.Errorf("%w: existing object %s differs", conflict, key)
	}
	return nil
}

func canonicalManifest(manifest dataArtifactManifest) ([]byte, error) {
	raw, err := json.Marshal(manifest)
	if err != nil {
		return nil, fmt.Errorf("marshal artifact manifest: %w", err)
	}
	encoded, err := canonical.CanonicalizeJSON(raw)
	if err != nil {
		return nil, fmt.Errorf("canonicalize artifact manifest: %w", err)
	}
	if err := validateManifestSchema(encoded); err != nil {
		return nil, fmt.Errorf("%w: artifact manifest: %v", ErrValidation, err)
	}
	return encoded, nil
}

func blobKey(hash string) (datastore.Key, error) {
	digest, ok := strings.CutPrefix(hash, "sha256:")
	if !ok {
		return datastore.Key{}, fmt.Errorf("invalid SHA-256 reference %q", hash)
	}
	return datastore.BlobKeySHA256Hex(digest)
}

func validateJSONSchema(schemaBytes, documentBytes []byte) error {
	schemaDocument, err := jsonschema.UnmarshalJSON(bytes.NewReader(schemaBytes))
	if err != nil {
		return fmt.Errorf("decode schema: %w", err)
	}
	instance, err := jsonschema.UnmarshalJSON(bytes.NewReader(documentBytes))
	if err != nil {
		return fmt.Errorf("decode document: %w", err)
	}
	compiler := jsonschema.NewCompiler()
	if err := compiler.AddResource("schema.json", schemaDocument); err != nil {
		return fmt.Errorf("register schema: %w", err)
	}
	compiled, err := compiler.Compile("schema.json")
	if err != nil {
		return fmt.Errorf("compile schema: %w", err)
	}
	return compiled.Validate(instance)
}

var manifestSchemaOnce sync.Once
var manifestSchema *jsonschema.Schema
var manifestSchemaErr error

func validateManifestSchema(document []byte) error {
	manifestSchemaOnce.Do(func() {
		schemaDocument, err := jsonschema.UnmarshalJSON(bytes.NewReader(schemacontract.DataArtifactManifestSchemaJSON))
		if err != nil {
			manifestSchemaErr = err
			return
		}
		compiler := jsonschema.NewCompiler()
		if err := compiler.AddResource(manifestSchemaRef, schemaDocument); err != nil {
			manifestSchemaErr = err
			return
		}
		manifestSchema, manifestSchemaErr = compiler.Compile(manifestSchemaRef)
	})
	if manifestSchemaErr != nil {
		return manifestSchemaErr
	}
	instance, err := jsonschema.UnmarshalJSON(bytes.NewReader(document))
	if err != nil {
		return err
	}
	return manifestSchema.Validate(instance)
}
